PLANNER-1324: Add a Shift Template Edit Form

Christopher-Chianelli opened the pull request:
Added an edit form similar to the edit form for the shifts in the shift roster for the shift templates in the rotation screen.

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/AbstractFormPopup.java:
```diff
@@ -0,0 +1,88 @@
+package org.optaweb.employeerostering.gwtui.client.common;
```
Missing license header.

Christopher-Chianelli replied:
Fixed

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalTimePicker.java:
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
```
It's 2019.

Christopher-Chianelli replied:
Fixed

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,242 @@
+                newDurationOfTemplate = Duration.ofDays(to.getRotationLength()).minus(newDurationFromReference).plusDays(to.getDayOffset()).plusSeconds(to.getTime().toSecondOfDay());
+            }
+            template.setDurationBetweenRotationStartAndTemplateStart(newDurationFromReference);
+            template.setShiftTemplateDuration(newDurationOfTemplate);
```
Lines 203-212 seem to be duplicate of lines 169-178. Could you extract the duplicate code into a method and reuse it? I believe it will improve readability as well.

Christopher-Chianelli replied:
Thanks; refactored

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    public void init(final ShiftTemplateGridObject shiftTemplateGridObject) {
+        this.shiftTemplateGridObject = (ShiftTemplateGridObject) shiftTemplateGridObject;
```
Redundant cast.

Christopher-Chianelli replied:
Thanks; removed.

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,242 @@
+    @Override
+    protected void onClose() {
+        if (shiftTemplateGridObject != null) {
+            shiftTemplateGridObject.getElement().classList.remove(SELECTED_CLASS);
```
I think that the code would be more concise and easier to read, test and maintain if the class manipulation was encapsulated in the `ShiftTemplateGridObject`. This could then be expressed as
```java
shiftTemplateGridObject.unselect();
// or
shiftTemplateGridObject.setSelected(false);
```

Christopher-Chianelli replied:
Will do

Christopher-Chianelli replied:
Done

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,242 @@
+import org.optaweb.employeerostering.shared.spot.Spot;
+import org.optaweb.employeerostering.shared.spot.SpotRestServiceBuilder;
+
+import static org.optaweb.employeerostering.gwtui.client.common.EventManager.Event.ROTATION_INVALIDATE;;
```
An extra semicolon.

Christopher-Chianelli replied:
Thanks; removed.

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,242 @@
+        showFor(shiftTemplateGridObject);
+    }
+
+    public void createNewShiftTemplate() {
```
This method seems to be unused. Can you confirm?

Christopher-Chianelli replied:
Confirmed; it a duplicate of the createNewShift() method. Might be useful to have later if we add a button to add a new Shift Template in the Rotation Screen, but as of now, unused.

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditForm.java:
```diff
@@ -50,94 +49,72 @@
+                         ListBox spotSelect, ListBox employeeSelect, HTMLInputElement pinned,
+                         HTMLButtonElement deleteButton, HTMLButtonElement applyButton,
+                         TenantStore tenantStore, EventManager eventManager,
+                         TranslationService translationService) {
```
Maybe one parameter per line?

Christopher-Chianelli replied:
It does read better (and makes for easier refactoring). Done.

Christopher-Chianelli commented:
@rsynek That actually the correct behavior; if TO day is less than FROM day, that means the shift "wrap around" the rotation. This happens for night shifts (ex: 9pm-6am) on the final day of rotation.

Christopher-Chianelli commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalTimePicker.java:
```diff
@@ -0,0 +1,126 @@
+        datePicker = BootstrapDateTimePicker.get(datePickerElement);
+        BootstrapDateTimePicker.BootstrapDateTimePickerOptions options = new BootstrapDateTimePicker.BootstrapDateTimePickerOptions();
+        options.format = DATE_FORMAT_STRING;
+        BootstrapDateTimePicker.BootstrapDateTimePickerIcons icons = new BootstrapDateTimePicker.BootstrapDateTimePickerIcons();
```
Fixed; removed.

Christopher-Chianelli commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,242 @@
+    private RotationTimeSelector to;
+
+    @DataField("spot")
+    private ListBox spotSelect; //FIXME: Don't use GWT widget
```
Done - the GWT widget does not have search capabilities, which mean you need to scroll through a list (of potentially hundreds employees)

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,262 @@
+        employeeSelect.addItem(translationService.format(I18nKeys.Shift_unassigned), "-1");
+
+        SpotRestServiceBuilder.getSpotList(tenantStore.getCurrentTenantId(), FailureShownRestCallback.onSuccess(spots -> {
+            spots.forEach(s -> this.spotSelect.addItem(s.getName(), s.getId().toString()));
```
The spot select box appears to never be enabled; in this case it doesn't make any sense to enumerate all the spots there - you need just one.

Opposed to ShiftEditForm, here is no "create button" that would use enabled select box.

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,262 @@
+        from.setTime(LocalTime.MIDNIGHT.plusSeconds(shiftTemplateView.getDurationBetweenReferenceAndStart().minusDays(startDayOffset).getSeconds()));
+        long endDayOffset = shiftTemplateView.getDurationBetweenReferenceAndStart().plus(shiftTemplateView.getShiftTemplateDuration()).toDays();
+        to.setDayOffset((int) endDayOffset);
+        to.setTime(LocalTime.MIDNIGHT.plusSeconds(shiftTemplateView.getDurationBetweenReferenceAndStart().plus(shiftTemplateView.getShiftTemplateDuration()).minusDays(endDayOffset).getSeconds()));
```
It's difficult to read such a long lines, please wrap at 120 chars.

rsynek commented:
> @rsynek That actually the correct behavior; if TO day is less than FROM day, that means the shift "wrap around" the rotation. This happens for night shifts (ex: 9pm-6am) on the final day of rotation.

Thanks for clarification, the final day of rotation is a good example. Question is whether it should behave like that even outside the final day of rotation. It may be difficult to prevent this behaviour since we don't know what the maximum shift duration is, right?

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/RotationTimeSelector.java:
```diff
@@ -0,0 +1,66 @@
+package org.optaweb.employeerostering.gwtui.client.pages.rotation;
```
License header.

Christopher-Chianelli replied:
Fixed

Christopher-Chianelli commented:
@rsynek Added a test for ShiftEditForm; created a utility class `CallbackFactory` so I can mock return values of Rest Methods. WDYT (I will add more tests)?

rsynek commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+    public void testInit() {
+        restCallbackAnswers.add(spotList);
+        restCallbackAnswers.add(employeeList);
+        ShiftView shiftView = new ShiftView(0, spotList.get(2),
```
The index (2) repeats through the test, let's create a constant with a descriptive name.

Christopher-Chianelli replied:
Done

rsynek commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+        restCallbackAnswers.add(spotList);
+        restCallbackAnswers.add(employeeList);
+        ShiftView shiftView = new ShiftView(0, spotList.get(2),
+                                            LocalDateTime.of(2012, 10, 1, 9, 0),
```
It seems the dates repeat pretty often, another candidate for a constant.

Christopher-Chianelli replied:
Done

rsynek commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+        testedShiftEditForm.init(mockShiftGridObject);
+
+        MouseEvent mouseEvent = mock(MouseEvent.class);
+        restCallbackAnswers.add(null);
```
Adding a null into a collection is very unusual. If there is not other way to do it, could you please at least put there a comment explaining the motivation?

Christopher-Chianelli replied:
Fixed; changed it into two different test cases; one for a successful delete (true) and another for a failed delete (false).

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/RotationTimeSelector.java:
```diff
@@ -0,0 +1,82 @@
+        offset = 0;
+        this.dayOffsetPicker = dayOffsetPicker;
+        this.timePicker = timePicker;
+        RosterRestServiceBuilder.getRosterState(tenantStore.getCurrentTenantId(), FailureShownRestCallback.onSuccess(this::setRotationLength));
```
Can you please explain why the `onSuccess` method of the `FailureShownRestCallback` has to be static? Is it some limitation of GWT/Errai?

Christopher-Chianelli replied:
`onSuccess` is a constructor for `FailureShownRestCallback`, so it doesn't make sense for it to be non-static. However, I created `CallbackFactory` so I can call  `FailureShownRestCallback.onSuccess` in an instance (and I can mock `CallbackFactory` so the callback is automatically called with mock return values).

Christopher-Chianelli commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+
+    @Before
+    public void setUp() throws Exception {
+        restCallbackAnswers = new LinkedList();
```
That List is unchecked; it need to contain `List<Spot>`, `List<Employee>` and other types returned by rest callbacks. I can change it to `LinkedList<Object>` but not anything more specific.

Christopher-Chianelli replied:
Fixed; Changed to `LinkedList<>`

Christopher-Chianelli commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+
+    private ShiftEditForm testedShiftEditForm;
+
+    private Queue<Object> restCallbackAnswers;
```
`setup` should reset theses (`restCallbackAnswers` will be set to a new empty linked list and `testedShiftEditForm` will be set to a new ShiftEditForm instance with the injected mocks).

yurloc commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,262 @@
+                    newDurationOfTemplate = Duration.ofDays(to.getRotationLength()).minus(newDurationFromReference).plusDays(to.getDayOffset()).plusSeconds(to.getTime().toSecondOfDay());
+                }
+                template.setDurationBetweenRotationStartAndTemplateStart(newDurationFromReference);
+                template.setShiftTemplateDuration(newDurationOfTemplate);
```
This (lines 194-203) is still duplicate of the body of `updateShiftTemplateFromWidgetsIfValid()`.

Christopher-Chianelli replied:
Fixed

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/CallbackFactory.java:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 201 Red Hat, Inc. and/or its affiliates.
```
Should have been 2019

Christopher-Chianelli replied:
Fixed

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditForm.java:
```diff
@@ -253,13 +203,40 @@ public void onApplyButtonClick(@ForEvent("click") final MouseEvent e) {
+            shiftView.setStartDateTime(from.getValue());
+            shiftView.setEndDateTime(to.getValue());
+            if (!shiftView.getStartDateTime().isBefore(shiftView.getEndDateTime())) {
+                throw new ValidationException("Shift's end time need to be after shift's start time.");
```
The `ValidationException` never gets to the user - this method catches it, returns `false`, which is never checked. As a result, the Shift returns to the old state and user gets no information about what went wrong.

The validation on individual fields is shown on the screen - it's probably compiled into the JS. Thus I am not sure if the ValidationException needs to be caught here.

Christopher-Chianelli replied:
The default error catcher doesn't work well here - it seems the `ValidationException` is converted to a Javascript exception (which in turn doesn't have a message). However, using NotificationFactory works as expected (although there the question of if we should allow the user to see the stack trace since this is an "expected" exception).

yurloc commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+                                            LocalDateTime.of(2012, 10, 1, 9, 0),
+                                            LocalDateTime.of(2012, 10, 1, 14, 0));
+        doReturn(shiftView).when(mockShiftGridObject).getShiftView();
+        doReturn(spotList.get(2)).when(mockShiftGridObject).getSpot();
```
Please prefer `when(...).thenReturn(...)` whenever possible.

See `doReturn()` Javadoc:
> Use `doReturn()` in those rare occasions when you cannot use `when(Object)`.
Beware that `when(Object)` is always recommended for stubbing because it is argument type-safe and more readable (especially when stubbing consecutive calls).

Christopher-Chianelli replied:
Done (and refactored all existing tests to use this pattern when possible)

yurloc commented:
The usage of static methods from REST request builders makes it difficult to mock the REST services the form depends on. This is a problem with the API generated by `gwt-jackson-rest-processor`. I have an idea how to improve this but the `restCallbackAnswers` queue is good workaround for the time being. Ideally, the form should avoid REST calls it doesn't need to do (`employeeList` and `spotList` data can be supplied from the outside) and require an `updateShiftViewRequest` instance so that it can be replaced with a mock in tests.

The testability could also be improved by making the form unaware of the grid object it is editing. It could take original `shiftView` data and require a success callback that would consume the updated `shiftView` data and use them to refresh the Shift grid object or create a new one.

Another possibility to improve the design for better testability (readability, maintenance, etc.) would be to apply the MVP pattern to separate the presentation and behavior concerns.

The `@Templated` class would be the View which only exposes the UI field accessors and invokes callbacks for Apply/Delete/Cancel buttons. It's very thin and there is not much to test.

The logic would go to a Presenter class. It uses View's field accessors to show data from the Model and get user input. It provides callbacks to act upon user actions (submit/cancel the from) and it talks to the Shift REST service to apply the changes and/or perform server-side data validation. It contains all kinds of decisions like "if the update was successful, hide the form and call the success callback provided by whoever requested the form to appear". The nice thing about MVP separation is that you can mock out the view when testing presenter and so you can use plain Mockito (not GwtMockito).

To sum it up, forms in Employee Rostering have a few design problems that make testing a bit more difficult than it needs to be. I think we should keep the scope of this PR limited and address complex refactoring separately. We should also discuss how to collaborate on this and how much effort it deserves. I believe that putting an extra effort into improving the code design will pay off by reducing test development and maintenance cost. We can't afford to write code without tests.

Christopher-Chianelli commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    private List<Spot> getSpotList() {
+        final String[] SPOT_NAMES = {"Spot1", "Spot2", "Spot3", "Spot4"};
```
Done

Christopher-Chianelli commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    private List<Employee> getEmployeeList() {
+        final String[] EMPLOYEE_NAMES = {"Employee1", "Employee2", "Employee3", "Employee4"};
```
Done

Christopher-Chianelli commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+        final String[] SPOT_NAMES = {"Spot1", "Spot2", "Spot3", "Spot4"};
+        List<Spot> out = new ArrayList<>(SPOT_NAMES.length);
+        for (int i = 0; i < SPOT_NAMES.length; i++) {
+            Spot spot = new Spot(0, SPOT_NAMES[i], Collections.emptySet());
```
Done

Christopher-Chianelli commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,387 @@
+        List<Spot> out = new ArrayList<>(SPOT_NAMES.length);
+        for (int i = 0; i < SPOT_NAMES.length; i++) {
+            Spot spot = new Spot(0, SPOT_NAMES[i], Collections.emptySet());
+            spot.setId((long) i);
```
Done

Christopher-Chianelli commented:
@rsynek @yurloc Integrated your suggested changes and added the ShiftTemplateEditForm tests. @yurloc I agree; refactoring the forms should be done in a different PR.

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalTimePicker.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Override
+    public void setValue(LocalTime value) {
```
@Christopher-Chianelli would it be possible to cover also this method by test(s)?

Christopher-Chianelli replied:
Nope; it uses JavaScript native methods, which make it impossible to unit test since it is impossible to mock those methods.

rsynek commented on employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java:
```diff
@@ -0,0 +1,248 @@
+            template.setRotationEmployeeId(parseId(employeeSelect.getSelectedValue()));
+            Duration newDurationFromReference = Duration.ofDays(from.getDayOffset()).plusSeconds(from.getTime().toSecondOfDay());
+            Duration newDurationOfTemplate;
+            if (to.getDayOffset() >= from.getDayOffset()) {
```
@Christopher-Chianelli if you move the logic of computing the `newDurationOfTemplate` into a new method, you can easily unit test it. Currently, the else branch is not covered and contains interesting logic.

Christopher-Chianelli replied:
Good idea; done

rsynek commented:
@Christopher-Chianelli thanks for adding more tests, we've got over 70% line and 40% branch coverage. I left there some hints about what else should be covered.

Regarding @yurloc suggestion about refactoring, I am fine with doing it in a new PR too. Although it might help you with writing tests, it's up to you.

Christopher-Chianelli commented:
@rsynek I found and fixed a bug involving drag; if you drag and your mouse over the shift template, it will cause the form to pop up. I have created a new class (`ClickableDecorator`) and added unit tests for it. I also refactored the methods `ShiftTemplateEditForm` for better unit testing.

yurloc commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,496 @@
+        verify(from).reportValidity();
+        verify(to).reportValidity();
+        verify(from).getValue();
+        verify(to).getValue();
```
All of these `verify()` statements are redundant as the calls to all of the method above are implicitly covered by the assertions below. Not a mistake, just a FYI and a suggestion how to make the test more concise.

yurloc commented on employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java:
```diff
@@ -0,0 +1,496 @@
+    private List<Employee> getEmployeeList() {
+        List<Employee> out = new ArrayList<>(NUM_OF_EMPLOYEES);
+        for (long i = 0; i < NUM_OF_EMPLOYEES; i++) {
+            Employee employee = new Employee(NUM_OF_SPOTS, "Employee" + i);
```
Typo I guess. Should be `new Employee(TENANT_ID, ...)`.

rsynek commented:
Thanks for the additional coverage. I agree with @yurloc that the remaining comments can be addressed in a separate PR. Did a quick manual check that the UI works, looks good.

Merging.

ge0ffrey commented:
Great to see that this PR is improved and merged on time for the cutoff!

@mramendi The issue has been merged, we hope this unblocks you.
